cunumeric/deferred.py

```diff
@@ -583,61 +586,79 @@ def _create_indexing_array(
             for i in range(key_store.ndim, rhs.ndim):
                 key_store = key_store.promote(i, rhs.shape[i])
 
-            out_dtype = rhs.dtype
-            # in the case this operation is called for the set_item, we
-            # return Point<N> type field that is later used for
-            # indirect copy operation
-            if is_set:
-                N = rhs.ndim
-                out_dtype = rhs.runtime.get_point_type(N)
-
-            # TODO : current implementation of the ND output regions
-            # requires out.ndim == rhs.ndim. This will be fixed in the
-            # future
-            out = rhs.runtime.create_unbound_thunk(out_dtype, ndim=rhs.ndim)
-            key_dims = key.ndim  # dimension of the original key
+            # has_set_value && set_value.size==1 corresponds to the case
+            # when a[bool_indices]=scalar
+            # then we can call "putmask" to modify input array
+            # and avoid calling Copy
+            has_set_value = set_value is not None and set_value.size == 1
+            if has_set_value:
+                mask = DeferredArray(
+                    self.runtime,
+                    base=key_store,
+                    dtype=self.dtype,
+                )
+                rhs.putmask(mask, set_value)
+                return False, rhs, rhs, self
```

Comment:
I'm not a big fan of this code. the name of the function is `_create_indexing_array`, but this code makes it do more than just creating indexing arrays. and in fact, the function was already complicated enough to warrant a refactoring, which I was thinking of doing at some point. I guess I'm fine with accepting this edit for this PR, but keep in mind that `_create_indexing_array` can use some refactoring for better maintainability.